File: binaryTrees/32.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inOrder(self, root: Optional[TreeNode], elems: list):
        if root is None:
            return

        self.inOrder(root.left, elems)

        if root:
            elems.append(root.val)

        self.inOrder(root.right, elems)

    @classmethod
    def getMinimumDifference(cls, root: Optional[TreeNode]) -> int:
        min_value = float("inf")
        elems = []

        cls().inOrder(root, elems)

        for index in range(1, len(elems)):
            min_value = min(min_value, abs(elems[index] - elems[index - 1]))

        return min_value


if __name__ == "__main__":
    root = TreeNode(4)
    root.left = TreeNode(2)
    root.right = TreeNode(6)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)

    print(Solution.getMinimumDifference(root))

File: binaryTrees/32.cpp
#include <bits/stdc++.h>
#include "../common/tree.h"

using namespace std;
using namespace Tree;


void inOrder(TreeNode* root, vector<int> &elems) {
    if(root == nullptr)
        return;

    inOrder(root->left, elems);

    if(root)
        elems.push_back(root->val);

    inOrder(root->right, elems);
}

int getMinimumDifference(TreeNode* root) {
    vector<int> elems;
    inOrder(root, elems);

    int min_value = INT_MAX;
    for(int i = 1; i < elems.size(); ++i)
        min_value = min(min_value, abs(elems[i] - elems[i - 1]));

    return min_value;
}



int main() {
    TreeNode* root = createNode(4);
    root->left=createNode(2);
    root->right=createNode(6);
    root->left->left=createNode(1);
    root->left->right=createNode(3);

    cout<<getMinimumDifference(root);
    
    // preOrder(root);
    // cout<<"\n";

    // inOrder(root);
    // cout<<"\n";
    // postOrder(root);
    // cout<<"\n";
    


    return 0;
}